### PR TITLE
Add flipping for proper visualization on wandb

### LINKED ISSUE #1272: WandB visualization shows flipped images even with no data augmentation
## Issue description
The images visualized on WandB (or, Tensorboard) appear flipped despite using no data augmentation method. When we look at the images on WandB (by setting `debugging=True` in the config file), we see that the images are flipped, thus leading us to think that we are feeding unrealistic inputs to the model and therefore modifying our training strategy (when we might have been right all along!). **Rather, the issue lies in how we input the matrix to torchvision's `make_grid` function that plots these images on WandB.** 

For concrete understanding, here's the config file for 2D model. Notice that there are no data augmentation methods used. 

<details><summary>config file</summary>
```console
{
    "command": "train",
    "gpu_ids": [3],
    "path_output": "outputs/2d-model_ccrop=192x80_res=1-1-1",
    "model_name": "modified_unet",
    "debugging": true,
    "object_detection_params": {
        "object_detection_path": null,
        "safety_factor": [1.0, 1.0, 1.0]
    },
    "wandb": {
        "wandb_api_key": "bf043c64b6c6b4abcc1ee7d8501c300b19945028",
        "project_name": "sci-zurich",
        "group_name": "ivadomed-2d-unet",
        "run_name": "ccrop=192x80_res=1-1-1",
        "log_grads_every": 1000
    },
    "loader_parameters": {
        "path_data": ["/home/GRAMES.POLYMTL.CA/u114716/duke/temp/muena/sci-zurich_preprocessed_20230119-161009"],
        "target_suffix": ["_lesion-manual"],
        "extensions": [".nii.gz"],
        "roi_params": {
            "suffix": null,
            "slice_filter_roi": null
        },
        "contrast_params": {
            "training_validation": ["T2w"],
            "testing": ["T2w"],
            "balance": {}
        },
        "slice_filter_params": {
            "filter_empty_mask": true,
            "filter_empty_input": true
        },
        "slice_axis": "sagittal",
        "multichannel": false,
        "soft_gt": false,
        "bids_validate": false
    },
    "split_dataset": {
        "fname_split": null,
        "random_seed": 6,
        "split_method" : "participant_id",
        "data_testing": {"data_type": null, "data_value":[]},
        "balance": null,
        "train_fraction": 0.75,
        "test_fraction": 0.2
    },
    "training_parameters": {
        "batch_size": 32,
        "loss": {
            "name": "DiceLoss"
        },
        "training_time": {
            "num_epochs": 200,
            "early_stopping_patience": 50,
            "early_stopping_epsilon": 0.001
        },
        "scheduler": {
            "initial_lr": 0.001,
            "lr_scheduler": {
                "name": "CosineAnnealingLR",
                "base_lr": 1e-5,
                "max_lr": 1e-2
            }
        },
        "balance_samples": {
            "applied": false,
            "type": "gt"
        },
        "mixup_alpha": null,
        "transfer_learning": {
            "retrain_model": null,
            "retrain_fraction": 1.0,
            "reset": true
        }
    },
    "default_model": {
        "name": "Unet",
        "dropout_rate": 0.25,
        "bn_momentum": 0.1,
        "final_activation": "relu",
        "depth": 4,
        "is_2d": true
    },
    "Modified3DUNet": {
        "applied": false,
        "length_3D": [128, 64, 48],
        "stride_3D": [128, 64, 48],
        "attention": false,
        "n_filters": 8
    },
    "uncertainty": {
        "epistemic": false,
        "aleatoric": false,
        "n_it": 0
    },
    "postprocessing": {
        "remove_noise": {"thr": -1},
        "keep_largest": {},
        "binarize_prediction": {"thr": 0.5},
        "uncertainty": {"thr": -1, "suffix": "_unc-vox.nii.gz"},
        "fill_holes": {},
        "remove_small": {"unit": "vox", "thr": 3}
    },
    "evaluation_parameters": {
        "target_size": {"unit": "vox", "thr": [20, 100]},
        "overlap": {"unit": "vox", "thr": 3}
    },
    "transformation": {
        "Resample":{
            "hspace": 1.0,
            "wspace": 1.0,
            "dspace": 1.0
        },
        "CenterCrop": {"size": [192, 80]},
        "NormalizeInstance": {"applied_to": ["im"]}
    }
}
```

</details>

Under the original code in [this function](https://github.com/ivadomed/ivadomed/blob/master/ivadomed/visualize.py#L132) in `visualize.py`, the images on WandB appear as below. Notice how the brainstem is at the bottom:

Screen Shot 2023-01-20 at 11 31 48 PM

Now, by using the _same config file_ but modifying the code so that input is `u-d` flipped before feeding as input to the `make_grid` function, that is, by doing:

```python
grid_img = vutils.make_grid(torch.flip(input_samples, [2,3]), normalize=True, scale_each=True)
```
in [this](https://github.com/ivadomed/ivadomed/blob/master/ivadomed/visualize.py#L174) line, we have the following images on WandB: 

Screen Shot 2023-01-20 at 11 38 54 PM

**NOTE:** We are not actually flipping the input before feeding it to the model, we are only flipping it for visualization purposes.

This suggests that there is a bug in how we visualize the images, hence leading to erroneous hypotheses about what the model might be seeing. 

This will be fixed in an imminent PR as it is important.

### PR DESCRIPTION
**Pull request by naga-karthik**
#### GitHub

- [x] I've given this PR a concise, self-descriptive, and meaningful title
- [x] I've linked relevant issues in the PR body
- [ ] I've applied [the relevant labels](https://www.neuro.polymtl.ca/software/contributing#pr_labels) to this PR
- [ ] I've assigned a reviewer

#### PR contents

- [ ] I've consulted [ivadomed's internal developer documentation](https://github.com/ivadomed/ivadomed/wiki) to ensure my contribution is in line with any relevant design decisions
- [ ] I've added [relevant tests](https://github.com/ivadomed/ivadomed/wiki/tests) for my contribution
- [ ] I've updated the [relevant documentation](https://github.com/ivadomed/ivadomed/wiki/documentation) for my changes, including argparse descriptions, docstrings, and ReadTheDocs tutorial pages

## Description

As mentioned in #1272, the image slices plotted on WandB (or, Tensorboard) appear flipped by default, even without using any data augmentation. This leads to incorrect understanding of what we think the model is seeing as its input and therefore, forcing us to change our training strategies when it is not required. 

This PR brings a minor change in the code snippet used for visualization, particularly, the images are now flipped in the [2,3] axes in order for them to appear correctly on WandB. 

## Linked issues

Resolves #1272

### CODE REVIEW

File: ivadomed/visualize.py [hunk 1]
```diff
@@ -171,21 +171,21 @@ def save_img(writer, epoch, dataset_type, input_samples, gt_samples, preds, wand
         elif isinstance(input_samples, list):
             input_samples = input_samples[0]
 
-        grid_img = vutils.make_grid(input_samples,
+        grid_img = vutils.make_grid(torch.flip(input_samples, [2,3]),
```
Comment (jcohenadad) [inline ivadomed/visualize.py:174]: Please add a comment (with reference to issue) to motivate this line of code

**Review by jcohenadad — COMMENTED** [review]
I'd like to make sure the propose change does not "break" with other modalities, eg histology. @hermancollin can you please verify? thanks

File: ivadomed/visualize.py (hunk 1 above)
Comment (naga-karthik) [reply ivadomed/visualize.py:178]: Great suggestion! Added it

Comment (jcohenadad) [reply ivadomed/visualize.py:178]: @naga-karthik when you address a comment, can you please always point to the commit? Eg:

```
Great suggestion! Added it in <commit number>
```

File: ivadomed/visualize.py [hunk 2]
```diff
@@ -171,21 +171,27 @@ def save_img(writer, epoch, dataset_type, input_samples, gt_samples, preds, wand
         elif isinstance(input_samples, list):
             input_samples = input_samples[0]
 
-        grid_img = vutils.make_grid(input_samples,
+        # the images fed to make_grid appear to be flipped by default even without any data augmentation, 
+        # possibly due to the orientation of the matrix. Hence, the images are flipped in the last two 
+        # dimensions in order for them to appear correctly in Tensorboard/WandB. For more details, see
+        # https://github.com/ivadomed/ivadomed/issues/1272
```
Comment (jcohenadad) [inline ivadomed/visualize.py:177]: Get to the point of the code right away-- start with the WHAT. And no need to write a novel-- details are already in the issue. 
```suggestion
        # Flipping the last two dimensions of the array to accommodate visual output in Tensorboard/WandB. 
        # https://github.com/ivadomed/ivadomed/issues/1272
```

Comment (naga-karthik) [reply ivadomed/visualize.py]: Yes, I'm working on trying to not write so many things and be more direct 😓 
Updated the comment in commit [904b143](https://github.com/ivadomed/ivadomed/pull/1273/commits/904b143e6d5f12b2f4dd5ec41900775803151319)

**Comment by naga-karthik** [comment]
Thank you for reviewing it @hermancollin ! Yes, it was 3D MRI data but trained using a 2D model. I thought that it would be the same for just 2D images as well. 

The option for checking if 2D data is being used is a good suggestion. Would you please be able to update the code in my branch?

**Comment by naga-karthik** [comment]
@hermancollin Thank you making the changes! However, I am still seeing inverted images for a 2D model I'm training. Here's the figure of inverted sagittal images 

Screen Shot 2023-01-30 at 12 34 33 PM

Is there any other way we could not use flipping only for microscopy images? Because as far as MRI data is concerned flipping has to be applied for both 2D and 3D models.

**Comment by naga-karthik** [comment]
@hermancollin I reverted back to the initial version of this fix in commit https://github.com/ivadomed/ivadomed/pull/1273/commits/b4229309beae62331634640185d8e32dd92d878d. Feel free to take a look if required, but, if not would you please approve it so that I can merge? Thanks!

**Review by hermancollin — APPROVED** [review]
To summarize the discussion, this change does introduce a flip for microscopy data but it doesn't really affect how we train our models for ADS.

For future reference, I was confused about the following argument
https://github.com/ivadomed/ivadomed/blob/b4229309beae62331634640185d8e32dd92d878d/ivadomed/visualize.py#L145

Even if the MRI data itself is 3D, the inputs for a 2DUnet model are obviously 2D so this variable could not be used to isolate microscopy data.
